File: Sample/ObjC/LoggingViewSample/ViewController.m
//
//  ViewController.m
//  LoggingViewSample
//
//  Created by Masaki Ando on 2018/12/12.
//  Copyright © 2018年 Hituzi Ando. All rights reserved.
//

#import <LoggingViewKit/LoggingViewKit-Swift.h>
#import <LoggingViewKit/LoggingViewKit.h>

#import "ViewController.h"

@interface ViewController () <LGVLoggingViewServiceDelegate>

@property (weak, nonatomic) IBOutlet LGVLabel *stepperLabel;
@property (weak, nonatomic) IBOutlet LGVView *sampleView;
@property (nonatomic) LGVButton *testButton;
@property (nonatomic) UIButton *getLogButton;

@end

@implementation ViewController

- (void) loadView {
    [super loadView];

    self.testButton = [[LGVButton alloc] initWithFrame:CGRectMake(0, 0, 100.f, 40.f)];
    self.testButton.loggingName = @"TestButton";
    self.testButton.logging = YES;
    self.testButton.touchableExtension = UIEdgeInsetsMake(20.f, 20.f, 20.f, 20.f);
    self.testButton.backgroundColor = [[UIColor redColor] colorWithAlphaComponent:.7f];
    [self.testButton setTitle:@"Test" forState:UIControlStateNormal];
    [self.testButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    [self.testButton addTarget:self
                        action:@selector(testButtonPressed:)
              forControlEvents:UIControlEventTouchUpInside];
    [self.sampleView addSubview:self.testButton];

    self.getLogButton = [[UIButton alloc] initWithFrame:CGRectMake(0, 44.f, 100.f, 40.f)];
    self.getLogButton.backgroundColor = [[UIColor greenColor] colorWithAlphaComponent:.7f];
    [self.getLogButton setTitle:@"Get Log" forState:UIControlStateNormal];
    [self.getLogButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    [self.getLogButton addTarget:self
                          action:@selector(getLogButtonPressed:)
                forControlEvents:UIControlEventTouchUpInside];
    [self.sampleView addSubview:self.getLogButton];
}

- (void) viewDidLoad {
    [super viewDidLoad];

    [LGVLoggingViewService sharedService].delegate = self;

    // Deletes all logs.
//    [[LGVLoggingViewService sharedService] deleteAllLogs];

    // Records viewDidLoad as the custom event.
    LGVLoggingAttribute *attr = [LGVLoggingAttribute attributeWithName:@"ViewController"];
    [[LGVLoggingViewService sharedService] customEvent:@"viewDidLoad" attribute:attr];

#ifdef DEBUG
    // Dumps the hierarchy of the root view.
    [LGVViewHierarchy dump:self.view];
#endif
}

#pragma mark - IBAction

- (IBAction) stepperChanged:(id)sender {
    LGVStepper *stepper = (LGVStepper *) sender;

    self.stepperLabel.text = [NSString stringWithFormat:@"%.1lf", stepper.value];
}

- (void) testButtonPressed:(LGVButton *)sender {
    NSLog(@"%@ Pressed", sender.loggingName);

    NSString *docDir = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES).firstObject;
    LGVFileDestination *destination = [LGVFileDestination destinationWithFile:@"debug.log"
                                                                  inDirectory:docDir
                                                                        error:NULL];
    [LVKLogViewController showFromViewController:self withSourceFile:destination];
}

- (void) getLogButtonPressed:(id)sender {
    // Records a click event.
    LGVLoggingAttribute *attr = [LGVLoggingAttribute attributeWithView:sender
                                                                  name:@"GetLogButton"
                                                        loggingEnabled:YES];

    attr.info = @{ @"more-info": @"test" };
    [[LGVLoggingViewService sharedService] click:attr];

    NSLog(@"All Logs: %@", [[LGVLoggingViewService sharedService] allLogs]);
    NSLog(@"Custom Event Logs: %@", [[LGVLoggingViewService sharedService] logsByEventType:@"viewDidLoad"]);
}

#pragma mark - LGVLoggingViewServiceDelegate

- (void) loggingViewService:(LGVLoggingViewService *)service
                willSaveLog:(LGVLog *)log
                  attribute:(LGVLoggingAttribute *)attribute {

}

- (void) loggingViewService:(LGVLoggingViewService *)service
                 didSaveLog:(LGVLog *)log
                  attribute:(LGVLoggingAttribute *)attribute
                      error:(LGVError *)error {

}

@end

File: Sample/Swift/LoggingViewSwiftSample/ViewController.swift
//
//  ViewController.swift
//  LoggingViewSwiftSample
//
//  Created by Masaki Ando on 2018/12/15.
//  Copyright © 2018年 Hituzi Ando. All rights reserved.
//

import LoggingViewKit
import UIKit

class ViewController: UIViewController, LGVLoggingViewServiceDelegate {
    @IBOutlet var stepperLabel: LGVLabel!
    @IBOutlet var sampleView: LGVView!

    private lazy var testButton: LGVButton = {
        let button = LGVButton(frame: CGRect(x: 0, y: 0, width: 100, height: 40))
        button.loggingName = "TestButton"
        button.isLogging = true
        button.touchableExtension = UIEdgeInsets(top: 20.0, left: 20.0, bottom: 20.0, right: 20.0)
        button.backgroundColor = UIColor.blue.withAlphaComponent(0.7)
        button.setTitle("Test", for: .normal)
        button.setTitleColor(UIColor.white, for: .normal)
        button.addTarget(self, action: #selector(testButtonPressed), for: .touchUpInside)
        return button
    }()

    private lazy var getLogButton: UIButton = {
        let button = UIButton(frame: CGRect(x: 0, y: 44, width: 100, height: 40))
        button.backgroundColor = UIColor.green.withAlphaComponent(0.7)
        button.setTitle("Get Log", for: .normal)
        button.setTitleColor(UIColor.white, for: .normal)
        button.addTarget(self, action: #selector(getLogButtonPressed), for: .touchUpInside)
        return button
    }()

    override func viewDidLoad() {
        super.viewDidLoad()

        sampleView.addSubview(testButton)
        sampleView.addSubview(getLogButton)

        LGVLoggingViewService.shared().delegate = self

        // Deletes all logs.
//        LGVLoggingViewService.shared().deleteAllLogs()

        // Records viewDidLoad as the custom event.
        let attr = LGVLoggingAttribute(name: "ViewController")
        LGVLoggingViewService.shared().customEvent("viewDidLoad", attribute: attr)

        #if DEBUG
        // Dumps hierarchy of the root view.
        LGVViewHierarchy.dump(view)
        #endif
    }

    // MARK: - IBAction

    @IBAction func stepperChanged(_ sender: Any) {
        if let stepper = sender as? LGVStepper {
            stepperLabel.text = String(format: "%.1lf", stepper.value)
        }
    }

    @objc func testButtonPressed(_ sender: LGVButton) {
        if let loggingName = sender.loggingName {
            print("\(loggingName) Pressed")
        }
    }

    @objc func getLogButtonPressed(_ sender: Any) {
        // Records a click event.
        let attr = LGVLoggingAttribute(view: sender,
                                       name: "GetLogButton",
                                       loggingEnabled: true)
        attr.info = ["more-info": "test"]
        LGVLoggingViewService.shared().click(attr)

        print("All Logs: \(LGVLoggingViewService.shared().allLogs())")
    }

    // MARK: - LGVLoggingViewServiceDelegate

    func loggingViewService(_ service: LGVLoggingViewService,
                            willSave log: LGVLog,
                            attribute: LGVLoggingAttribute) {}

    func loggingViewService(_ service: LGVLoggingViewService,
                            didSave log: LGVLog,
                            attribute: LGVLoggingAttribute,
                            error: LGVError?) {}
}
